File: src/main/java/buy_coffee/scala_way/Cafe.scala
package buy_coffee.scala_way

class Cafe {
  def buyCoffee(cc: CreditCard): (Coffee, Charge) = {
    val cup = Coffee(5)
    (cup, Charge(cc, cup.price))
  }

  def buyCoffee(cc: CreditCard, n: Int): (List[Coffee], Charge) = {
    val purchases: List[(Coffee, Charge)] = List.fill(n) (buyCoffee(cc))
    val (coffees, charges) = purchases.unzip
    (coffees, charges.reduce((c1, c2) => c1.combine(c2)))
  }

  def coalesce(charges: List[Charge]): List[Charge] = {
    charges.groupBy(_.cc).values.map(_.reduce(_ combine _)).toList
  }
}

case class Coffee(price: Double) {
}

case class CreditCard(name: String) {
}

case class Charge(cc: CreditCard, amount: Double) {

  def combine(other: Charge): Charge = {
    if (cc == other.cc) {
      Charge(cc, amount + other.amount)
    } else {
      throw new Exception("can't combine charges to different cards")
    }
  }
}

object Cafe {
  def main(args: Array[String]): Unit = {
    val c = new Cafe()
    val cofe1 = c.buyCoffee(CreditCard("wyl"))
    val cofe2 = c.buyCoffee(CreditCard("wyl"), 2)
    val cofe3 = c.buyCoffee(CreditCard("ramble"), 3)

    val charges = c.coalesce(List[Charge](cofe1._2, cofe2._2, cofe3._2))

    println(cofe1)
    println(cofe2)
    println(cofe3)
    println(charges)
  }
}

File: src/main/java/buy_coffee/java_way/Cafe.java
package buy_coffee.java_way;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;

public class Cafe {

    public Account buyCoffee(CreditCard cc) {
        Coffee cup = new Coffee(5);
        return new Account(cup, new Charge(cc, cup.price));
    }

    public Accounts buyCoffee(CreditCard cc, int n) {
        List<Coffee> coffees = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            coffees.add(new Coffee(5));
        }
        return new Accounts(coffees, new Charge(cc, coffees.stream().mapToDouble(c -> c.price).sum()));
    }

    public List<Charge> coalesce(List<Charge> charges) {
         return charges.stream().collect(Collectors.groupingBy(c -> c.creditCard.name)).values().stream()
                 .map(c -> c.stream().reduce(Charge::combine).orElse(null)).collect(Collectors.toList());
    }

    public static void main(String[] args) {
        Cafe c = new Cafe();
        Account cofe1 = c.buyCoffee(new CreditCard("wyl"));
        Accounts cofe2 = c.buyCoffee(new CreditCard("wyl"), 2);
        Accounts cofe3 = c.buyCoffee(new CreditCard("ramble"), 5);

        List<Charge> charges = c.coalesce(Arrays.asList(cofe1.charge, cofe2.charge, cofe3.charge));

        System.out.println(cofe1);
        System.out.println(cofe2);
        System.out.println(cofe3);
        System.out.println(charges);
    }
}
